### PR TITLE
[FIX] *: Remove unused static files

### PR DESCRIPTION
**Pull request by jaeschwa**
Added a test to check all the static files that are in use or not. 
Update all the modules so that files that are not in use are removed.
The test works as : 
- Retrieving all files in static folder in a set
- Remove the occurrence of these files from the set
- If the set is not empty at the end files are not used.

task-4001616

### CODE REVIEW

**Comment by jaeschwa** [comment]
I will rename the commit just after!

**Review by vava-odoo — CHANGES_REQUESTED** [review]
Great cleaning work. Though I fear you must slightly adapt your test to avoid removing files referenced in the website.

File: tests/test_generic/tests/test_xml.py [hunk 1]
```diff
@@ -46,18 +50,35 @@ def _check_files_in_path(self, module):
...
+        tree = etree.fromstring(content.encode('utf8'))
+        for element in tree.iter():
+            file_name = element.attrib.get('file', [])
+            if 'static/src/binary' in file_name and file_name in static_files:
+                static_files.remove(file_name)
+            if element.text is not None:
+                file_names = re.findall(r'src="([^"]+)"', element.text)
+                for file_name in file_names:
+                    if file_name in static_files:
+                        static_files.remove(file_name)
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:81]: No it is not because the same file can appear multiple times. So it will trigger an error.

File: tests/test_generic/tests/test_xml.py [hunk 2]
```diff
@@ -27,10 +27,14 @@ def test_xml_files(self):
     def _check_files_in_path(self, module):
         path = get_industry_path() + module
         is_studio_required = False
+        static_files = set()
         for root, dirs, files in os.walk(path):
+            dirs.sort(reverse=True)
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:32]: I am on it!

File: tests/test_generic/tests/test_xml.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vava-odoo) [reply tests/test_generic/tests/test_xml.py:81]: Ah yes, the second one is a list. What about this then?
```suggestion
                static_files -= set(file_names)
```

File: tests/test_generic/tests/test_xml.py [hunk 3]
```diff
@@ -46,18 +53,35 @@ def _check_files_in_path(self, module):
...
                 self._check_knowledge_article_is_published(content, file_name)
                 self._check_duplicate_records(content, file_name)
                 self._check_website_published_false(module, file_name)
+                self._check_static_files_usage_in_xml(content, in_use_files)
                 if root.split('/')[-1] == 'data':
                     self._check_is_published_false(module, file_name)
                     if not is_studio_required:
                         is_studio_required = self._check_studio(content, file_name)
         self._check_manifest(manifest_content, is_studio_required)
+        in_use_files = set(file.lstrip('/') for file in in_use_files)
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:70]: Don't you need to remove the name of the module as well?

File: tests/test_generic/tests/test_xml.py [hunk 4]
```diff
@@ -46,18 +53,35 @@ def _check_files_in_path(self, module):
...
+        for file in static_files - in_use_files:
+            _logger.warning(f"Unused static file {file}.")
+        for file in in_use_files - static_files:
+            _logger.warning(f"No reference found for this file: {file}.")
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:74]: This would raise an error, no? But yeah, keep it.

File: tests/test_generic/tests/test_xml.py [hunk 5]
```diff
@@ -27,10 +27,17 @@ def test_xml_files(self):
     def _check_files_in_path(self, module):
         path = get_industry_path() + module
         is_studio_required = False
+        static_files = set()
+        in_use_files = set()
         for root, dirs, files in os.walk(path):
+            dirs.sort(
+                reverse=True
+            )  # sort the directory by alphabetical order so static directory is read first.
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:35]:
```suggestion
            # sort the directory by alphabetical order so static directory is read first.
            dirs.sort(reverse=True)
```

File: tests/test_generic/tests/test_xml.py [hunk 6]
```diff
@@ -27,10 +27,17 @@ def test_xml_files(self):
...
             for file_name in files:
                 file_path = os.path.join(root, file_name)
                 ext = os.path.splitext(file_path)[1].lower()
+                if 'static/src/binary' in file_path:
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:39]: I don't know exactly how to proceed, but there are more paths to check. See for instance photography, with static/demo/img ...

File: tests/test_generic/tests/test_industry_requirements.py [hunk 7]
```diff
@@ -77,7 +77,7 @@ def test_cloc_exclude_view(self):
     def test_sale_ok_and_is_published_in_db(self):
         models = ["product.template", "product.product"]
         for model in models:
-            if "is_published" in self.env[model]._fields:
+            if model in self.env and "is_published" in self.env[model]._fields:
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_industry_requirements.py:80]: which module doesn't use product? :thinking:

File: tests/test_generic/tests/test_xml.py [hunk 8]
```diff
@@ -76,12 +76,14 @@ def _check_files_in_path(self, module):
...
-                static_files.add(file_name)
+            file_names = set(
+                element.attrib.get(key) for key in ['file', 'src', 'data-original-src']
+            )
             if element.text is not None:
-                file_names = re.findall(r'src="(?!/web/)([^"]+)"', element.text)
-                static_files.update(file_names)
+                file_names.update(re.findall(r'src="(?!/web/)([^"]+)"', element.text))
+
+            file_names.remove(None)
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py]: why do you need that?

File: clothing_boutique/demo/website_page_views.xml [hunk 9]
```diff
@@ -372,7 +372,7 @@ to Saturday&amp;nbsp;at 10 Am - 7:30 PM</font>
                                     <div class="col-lg-3 offset-lg-1 o_colored_level" style="">
                                         <p class="text-muted"><b>The Glam Boutique</b><br/>250 Executive Park Blvd, Suite 3400 <br/> San Francisco CA 94134 <br/>United States</p>
                                         <p class="text-muted">
-                                            <img src="/clothing_boutique/static/src/binary/ir_attachment/940-glamlogo.png" alt="" class="img img-fluid o_we_custom_image me-auto float-start" data-original-id="939" data-original-src="/web/image/939-36b131e2/glam%20logo.png" data-mimetype="image/png" data-bs-original-title="" aria-describedby="tooltip126360" style="width: 100% !important;" loading="lazy"/>
+                                            <img src="/clothing_boutique/static/src/binary/ir_attachment/939-glamlogo.png" alt="" class="img img-fluid o_we_custom_image me-auto float-start" data-original-id="939" data-original-src="/web/image/939-36b131e2/glam%20logo.png" data-mimetype="image/png" data-bs-original-title="" aria-describedby="tooltip126360" style="width: 100% !important;" loading="lazy"/>
```
Comment (vava-odoo) [inline clothing_boutique/demo/website_page_views.xml:375]: Why this change?

**Review by vava-odoo — COMMENTED** [review]
Please also have a look at unresolved issues

File: clothing_boutique/demo/website_page_views.xml (hunk 9 above)
Comment (jaeschwa) [reply clothing_boutique/demo/website_page_views.xml:375]: File 939 does not exist, and since data-original-id="939", I guessed it was a typo error.

File: tests/test_generic/tests/test_xml.py (hunk 3 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:69]: The name of the module is in the static files list.

File: tests/test_generic/tests/test_xml.py (hunk 4 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:73]: Well I tought but no since I deleted files in a previous commit and all tests passed.

File: tests/test_generic/tests/test_xml.py (hunk 6 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py]: Taken into account

File: tests/test_generic/tests/test_industry_requirements.py (hunk 7 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_industry_requirements.py:80]: I do not remember but this raised an error.

File: tests/test_generic/tests/test_xml.py (hunk 8 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py]: I changed it!

File: tests/test_generic/tests/test_industry_requirements.py (hunk 7 above)
Comment (vava-odoo) [reply tests/test_generic/tests/test_industry_requirements.py:80]: Really weird. Could you push again without this change to show the error?

File: industry_lawyer/static/src/binary/website/1-logo
[the remark this answers is not in the document]
Comment (jaeschwa) [reply industry_lawyer/static/src/binary/website/1-logo:1]: It was not used, but I fixed it

File: tests/test_generic/tests/test_xml.py [hunk 10]
```diff
@@ -46,18 +52,46 @@ def _check_files_in_path(self, module):
...
+                if element.get("name") == "cover_properties":
+                    file_names.update(re.findall(r"url\(['\"]?([^'\")]+)['\"]?\)", element.text))
+                file_names.update(re.findall(r'src="([^"]+)"', element.text))
+
+            in_use_files.update(
+                {
+                    file
+                    for file in file_names
+                    if file
+                    and not file.startswith(('web', '/web', 'https', '/unsplash', 'sale_timesheet'))
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:92]: I'd add the file in static of the industry instead of escaping the test here
```suggestion
                    and not file.startswith(('web', '/web', 'https', '/unsplash'))
```

File: tests/test_generic/tests/test_xml.py [hunk 11]
```diff
@@ -46,18 +52,46 @@ def _check_files_in_path(self, module):
...
                 self._check_knowledge_article_is_published(content, file_name)
                 self._check_duplicate_records(content, file_name)
                 self._check_website_published_false(module, file_name)
+                self._check_static_files_usage_in_xml(content, in_use_files)
                 if root.split('/')[-1] == 'data':
                     self._check_is_published_false(module, file_name)
                     if not is_studio_required:
                         is_studio_required = self._check_studio(content, file_name)
         self._check_manifest(manifest_content, is_studio_required)
+        print(in_use_files)
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:69]: :eyes:

File: headhunter/demo/website_view.xml [hunk 12]
```diff
@@ -67,7 +67,7 @@
                             <div class="container-fluid">
                                 <div class="row o_grid_mode" data-row-count="10" style="">
                                     <div class="o_grid_item col-lg-6 g-col-lg-6 g-height-10 o_grid_item_image text-center o_colored_level" data-name="Block" style="grid-area: 1 / 1 / 11 / 7; z-index: 1; background-image: none;">
-                                        <img src="/web/image/website.s_masonry_block_default_image_1" class="img img-fluid mx-auto" alt="" loading="lazy" data-original-id="363" data-original-src="/theme_buzzy/static/src/img/photos/s_masonry_block_default_image_1.jpg" data-mimetype="image/jpeg" data-bs-original-title="" title="" aria-describedby="tooltip406637" style=""/>
+                                        <img src="/web/image/website.s_masonry_block_default_image_1" class="img img-fluid mx-auto" alt="" loading="lazy" data-mimetype="image/jpeg" data-bs-original-title="" title="" aria-describedby="tooltip406637" style=""/>
```
Comment (vava-odoo) [inline headhunter/demo/website_view.xml:70]: Useless diff?

**Review by vava-odoo — APPROVED** [review]
Only minor comments, thanks
